python/lsst/daf/butler/core/datasets/type.py

```diff
@@ -131,6 +140,28 @@ def __init__(self, name: str, dimensions: Union[DimensionGraph, Iterable[Dimensi
             self._storageClass = None
             self._storageClassName = storageClass
 
+        self._parentStorageClass: Optional[StorageClass] = None
+        self._parentStorageClassName: Optional[str] = None
+        if parentStorageClass is not None:
+            # Only allowed for a component dataset type
+            _, componentName = self.splitDatasetTypeName(self._name)
+            if componentName is None:
+                raise ValueError("Can not specify a parent storage class if this is not a component"
+                                 f" ({self._name})")
+            if isinstance(parentStorageClass, StorageClass):
+                self._parentStorageClass = parentStorageClass
+                self._parentStorageClassName = parentStorageClass.name
+            else:
+                self._parentStorageClassName = parentStorageClass
+
+        # Temporarily warn if this is a component but is missing the
+        # parent storage class. Usually implies some place not using the right
+        # API.
+        _, componentName = self.splitDatasetTypeName(self._name)
+        if parentStorageClass is None and componentName is not None:
+            log.warning("Component dataset type '%s' constructed without parent storage class",
+                        self._name)
```

Comment:
Unless this fails tests that are a pain to fix, I'd be happy to bump this up to an exception.

Reply:
I set it to be an exception to find where tests were breaking. I think a couple of tests were still morning (mostly the ones that set up fake datasetrefs). -- I'll take another look at them. Raising an exception will mean that this code will fail if an old registry is used when the code encounters the component dataset types but I imagine that's probably okay.